src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp

```diff
@@ -0,0 +1,125 @@
+
+#include "ai/hl/stp/action/chip_action.h"
+#include "ai/hl/stp/action/move_action.h"
+#include "ai/hl/stp/evaluation/calc_best_shot.h"
+#include "geom/point.h"
+#include "geom/ray.h"
+#include "geom/segment.h"
+#include "geom/util.h"
+#include "shared/constants.h"
+
+GoalieTactic::GoalieTactic(const Ball &ball, const Field &field,
+                           const Team &friendly_team, const Team &enemy_team)
+    : ball(ball),
+      field(field),
+      friendly_team(friendly_team),
+      enemy_team(enemy_team),
+      Tactic(true)
+{
+}
+
+std::string GoalieTactic::getName() const
+{
+    return "Goalie Tactic";
+}
+
+void GoalieTactic::updateParams(
+    const Ball &ball, const Field &field, const Team &friendly_team,
+    const Team &enemy_team, const std::optional<Evaluation::EnemyThreat> &enemy_threat)
+{
+    // Update the parameters stored by this Tactic
+    this->ball          = ball;
+    this->field         = field;
+    this->friendly_team = friendly_team;
+    this->enemy_team    = enemy_team;
+    this->enemy_threat  = enemy_threat;
+}
+
+double GoalieTactic::calculateRobotCost(const Robot &robot, const World &world)
+{
+    // Prefer robots closer to the destination
+    // We normalize with the total field length so that robots that are within the field
+    // have a cost less than 1
+    double cost = (robot.position() - world.field().friendlyGoal()).len() /
+                  world.field().totalLength();
+    return std::clamp<double>(cost, 0, 1);
+}
+
+void GoalieTactic::calculateNextIntent(IntentCoroutine::push_type &yield)
+{
+    MoveAction move_action = MoveAction();
+    ChipAction chip_action = ChipAction();
+    do
+    {
+        Ray ball_ray = Ray(ball.position(), ball.velocity());
+
+        // Create a segment along the goal line, slightly shortened to account for the
+        // robot radius so as we move along the segment we don't try run into the goal
+        // posts
+        Point p1                        = Point(field.friendlyGoalpostNeg().x(),
+                         field.friendlyGoalpostNeg().y() + ROBOT_MAX_RADIUS_METERS);
+        Point p2                        = Point(field.friendlyGoalpostPos().x(),
+                         field.friendlyGoalpostPos().y() - ROBOT_MAX_RADIUS_METERS);
+        Segment goalie_movement_segment = Segment(p1, p2);
+        Segment full_goal_segment =
+            Segment(field.friendlyGoalpostNeg(), field.friendlyGoalpostPos());
+
+        // Our default behavior is to track the ball along the goal line, and always face
+        // the ball
+        Point goalie_pos = closestPointOnSeg(ball.position(), goalie_movement_segment);
+        Angle goalie_orientation = (ball.position() - goalie_pos).orientation();
+        std::unique_ptr<Intent> next_intent;
+
+        auto [intersection1, intersection2] =
+            raySegmentIntersection(ball_ray, full_goal_segment);
+        if (intersection1 && ball.velocity().len() > 0.2)
```

Comment:
Nit;
I would like to see intersection1.has_value() 

I think that's a lot more clear as you're using auto here.

Reply:
fixed